Update readme

thaven opened the pull request.

thaven committed:
Minor text improvements

thaven committed:
Mention struct support and update example

thaven committed:
Add overview of supported types

WebFreak001 commented on Readme.md:
```diff
@@ -65,8 +65,8 @@ Setting read only properties results in a thrown `DBusException`.
 After that you can request a name so that other DBus clients can connect to your program.
 
-You can return a `Tuple!(args)` to return multiple values (multiple out values in XML) or
-return a `Variant!DBusAny` to support returning any dynamic value.
```
If you wrap this you should also wrap the next change

thaven replied:
Thinking it over once again... I made a mistake here. Returning a struct is possible, but in this particular case, it will not elicit the same behavior as returning a `Tuple`. I will revert this change.

WebFreak001 commented on Readme.md:
```diff
@@ -108,37 +108,64 @@ See the Concurrent Updates section for details how to implement this in a custom
+All DBus-compatible basic types work (except file descriptors).
+Any forward range can be marshaled in as DBus array of that type but arrays must be taken out as dynamic arrays.
+
+As per version 2.3.0, D `struct` types are fully supported by `ddbus`. By default all public fields of a structure are marshaled. This behavior can be changed by UDAs. Mapping DBus structures to a matching instance of `std.typecons.Tuple`, like earlier versions of `ddbus` did, is also still supported.
```
can you explain how it is still supported?

thaven replied:
Left this as is, but added some more explanation to the type mapping table below.

thaven committed:
Reverted erroneous change

thaven committed:
Clarify mapping of types

thaven committed:
Explain customization of struct marshaling

WebFreak001 commented on Readme.md:
```diff
@@ -108,37 +108,92 @@ See the Concurrent Updates section for details how to implement this in a custom
+| `Tuple!(T...)` | structure | The DBus structure will map all of the `Tuple`'s values in sequence. |
+| any `struct` | structure | The DBus structure will map all public fields of the `struct` type in order of definition, unless otherwise specified using UDAs. |
+| `ddbus` style variant `Variant!T` | variant | `Variant!T` is in fact just a wrapper type to force representation as a variant in DBus, use `Variant!DBusAny` for actual dynamic typing. |
+| Phobos style variants `std.variant.VariantN` | variant | Only supported if set of allowed types is limited to types that can be marshaled by `ddbus`, so `std.variant.Variant` is not supported, but `std.variant.Algebraic` may be, depending on allowed types |
```
Uh I know this will be nicely formatted by github but for people who cloned the repository and now look at the README in vim or some other text editor this will be really unreadable, can you add spaces in the entire table to make it look like a table in text mode

WebFreak001 commented on Readme.md:
````diff
@@ -108,37 +108,92 @@ See the Concurrent Updates section for details how to implement this in a custom
+will appear freshly initialized. This is true even when just converting a `struct` to
+`DBusAny` and back.
+
+```D
````
in the other code block this is a lowercase d

WebFreak001 commented on Readme.md:
```diff
@@ -108,37 +108,92 @@ See the Concurrent Updates section for details how to implement this in a custom
 
-Importing `ddbus` publicly imports the `thin`,`router`,`bus` and `simple` modules.
-These provide most of the functionality you probably want,
+Importing `ddbus` will publicly import the `thin`,`router`,`bus`, `simple` and
```
you added a space between `bus` and `simple` but not the other values

thaven committed:
Applied requested changes

- changed language indicator on code block to lower case
- added spaces to make table look like a table in plain text. removed final `|` to avoid lots of spaces at the end of most lines (has no effect on markdown rendering).
- added spaces after commas in list of modules imported through package.d

WebFreak001 commented:
hm not sure about leaving out the last | characters of the table in each row because of different markdown implementations like the one on the dub registry

thaven commented:
Looks like that should be OK at least for dub registry, which uses the markdown parser of vibe.d.

I checked the source, it always strips off \| at the beginning and end of the line if those happen to be there.

WebFreak001 reviewed (approved):
I think the "As per version" block should mention the uda needed for that because of TL;DR people but otherwise it looks good

thaven committed:
Turn mention of UDAs into a link to their description

thaven committed:
Update version number in dub.json example
